File: leetcode/c++/easy/_1790_check_if_one_string_swap_can_make_strings_equal.cpp
// Leetcode Problem 1790

#include <gtest/gtest.h>

using namespace std;

class Solution {
public:
    bool areAlmostEqual(string s1, string s2) {
        if (s1 == s2) {
            return true;
        }

        int pos1 = -1;
        int pos2 = -1;

        for (int i = 0; i < s1.size(); i++) {
            if (s1[i] != s2[i]) {
                if (pos1 == -1) {
                    pos1 = i;
                } else if (pos2 == -1) {
                    pos2 = i;
                } else {
                    return false;
                }
            }
        }

        if (pos1 != -1 && pos2 == -1) {
            return false;
        }

        return ((s1[pos1] == s2[pos2]) && (s1[pos2] == s2[pos1]));
    }
};


class TestAreAlmostEqual : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestAreAlmostEqual, SimpleTestCase1) {
    string s1 = "bank";
    string s2 = "kanb";

    EXPECT_TRUE(solution.areAlmostEqual(s1, s2));
}

TEST_F(TestAreAlmostEqual, SimpleTestCase2) {
    string s1 = "attack";
    string s2 = "defense";

    EXPECT_FALSE(solution.areAlmostEqual(s1, s2));
}

TEST_F(TestAreAlmostEqual, SimpleTestCase3) {
    string s1 = "kelb";
    string s2 = "kelb";

    EXPECT_TRUE(solution.areAlmostEqual(s1, s2));
}

TEST_F(TestAreAlmostEqual, SimpleTestCase4) {
    string s1 = "ahusgbejsh";
    string s2 = "ahusgbjesh";

    EXPECT_TRUE(solution.areAlmostEqual(s1, s2));
}

TEST_F(TestAreAlmostEqual, SimpleTestCase5) {
    string s1 = "a";
    string s2 = "b";

    EXPECT_FALSE(solution.areAlmostEqual(s1, s2));
}

File: leetcode/rust/problems/src/easy/_1790_check_if_one_string_swap_can_make_strings_equal.rs
// Leetcode Problem 1790

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn are_almost_equal(s1: String, s2: String) -> bool {
        let mut pos1: i32 = -1;
        let mut pos2: i32 = -1;

        for i in 0..s1.len() {
            if s1.chars().nth(i) != s2.chars().nth(i) {
                if pos1 == -1 {
                    pos1 = i as i32;
                } else if pos2 == -1 {
                    pos2 = i as i32;
                } else {
                    return false;
                }
            }
        }

        if pos1 != -1 && pos2 == -1 {
            return false;
        }

        (s1.chars().nth(pos1 as usize) == s2.chars().nth(pos2 as usize)) && (s2.chars().nth(pos1 as usize) == s1.chars().nth(pos2 as usize))
    }
}


#[test]
fn test_are_almost_equal() {
    let s11 = String::from("bank");
    let s12 = String::from("kanb");

    let s21 = String::from("attack");
    let s22 = String::from("defend");

    let s31 = String::from("kelb");
    let s32 = String::from("kelb");

    let s41 = String::from("ahusgbejsh");
    let s42 = String::from("ahusgbjesh");

    let s51 = String::from("a");
    let s52 = String::from("b");

    assert!(Solution::are_almost_equal(s11, s12));
    assert!(!Solution::are_almost_equal(s21, s22));
    assert!(Solution::are_almost_equal(s31, s32));
    assert!(Solution::are_almost_equal(s41, s42));
    assert!(!Solution::are_almost_equal(s51, s52));
}